direct/io-bulk-fs/src/main/java/cz/o2/proxima/direct/bulk/DefaultNamingConvention.java

```diff
@@ -0,0 +1,124 @@
+/**
+ * Copyright 2017-2020 O2 Czech Republic, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.direct.bulk;
+
+import com.google.common.annotations.VisibleForTesting;
+import cz.o2.proxima.annotations.Internal;
+import cz.o2.proxima.functional.Factory;
+import cz.o2.proxima.util.Pair;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.UUID;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@Internal
+public class DefaultNamingConvention implements NamingConvention {
+
+  private static final char SEPARATOR = '/';
+  private static final Pattern NAME_PATTERN =
+      Pattern.compile(".*/?[^/]+-([0-9]+)_([0-9]+)[^/]*\\.+[^/]*$");
+  private static final DateTimeFormatter DIR_FORMAT = DateTimeFormatter.ofPattern("yyyy/MM/");
+
+  private final long rollPeriodMs;
+  private final String prefix;
+  private final String suffix;
+  private final Factory<String> uuidGenerator;
+
+  DefaultNamingConvention(Duration rollPeriod, String prefix, String suffix) {
+    this(rollPeriod, prefix, suffix, () -> UUID.randomUUID().toString());
+  }
+
+  @VisibleForTesting
+  public DefaultNamingConvention(
+      Duration rollPeriod, String prefix, String suffix, Factory<String> uuidGenerator) {
+    this.rollPeriodMs = rollPeriod.toMillis();
+    this.prefix = prefix;
+    this.suffix = suffix;
+    this.uuidGenerator = uuidGenerator;
+  }
+
+  @Override
+  public String nameOf(long ts) {
+    long boundary = ts - ts % rollPeriodMs;
+    String date =
+        DIR_FORMAT.format(
+            LocalDateTime.ofInstant(
+                Instant.ofEpochMilli(boundary), ZoneId.ofOffset("UTC", ZoneOffset.UTC)));
+    return String.format(
+        "/%s%s-%d_%d_%s.%s",
+        date, prefix, boundary, boundary + rollPeriodMs, uuidGenerator.apply(), suffix);
+  }
+
+  @Override
+  public Set<String> prefixesOf(long startStamp, long endStamp) {
```

Comment:
Would it make sense to change the API to work with Instants instead? (That applies to FileSystem abstraction too). (these are easier to work with and most of the methods I see convert timestamps to Instants anyway)

Reply:
This is actually a big topic. :-)
I think `Instant` and `Duration` should be treated in similar way to `Optional` - i.e. use it in public APIs, convert to raw types in internal APIs. This is internal API and I would keep it as is for now. We can reiterate on this later.